executor/analyze.go

```diff
@@ -78,40 +78,48 @@ func (e *AnalyzeExec) Next() (*Row, error) {
 		return nil, errors.Trace(err)
 	}
 	taskCh := make(chan *analyzeTask, len(e.tasks))
-	resultCh := make(chan analyzeResult, len(e.tasks))
+	resultCh := make(chan statistics.AnalyzeResult, len(e.tasks))
 	for i := 0; i < concurrency; i++ {
 		go e.analyzeWorker(taskCh, resultCh)
 	}
 	for _, task := range e.tasks {
 		taskCh <- task
 	}
 	close(taskCh)
-	results := make([]analyzeResult, 0, len(e.tasks))
+	dom := sessionctx.GetDomain(e.ctx)
+	lease := dom.StatsHandle().Lease
+	if lease > 0 {
+		for i := 0; i < len(e.tasks); i++ {
+			result := <-resultCh
+			if result.Err != nil {
+				return nil, errors.Trace(err)
+			}
+			result.Ctx = e.ctx
+			dom.StatsHandle().AnalyzeResultCh() <- &result
+		}
+		// We sleep two lease to make sure other tidb node has updated this node.
+		time.Sleep(lease * 2)
+		return nil, nil
+	}
+	results := make([]statistics.AnalyzeResult, 0, len(e.tasks))
 	for i := 0; i < len(e.tasks); i++ {
 		result := <-resultCh
 		results = append(results, result)
-		if result.err != nil {
+		if result.Err != nil {
 			return nil, errors.Trace(err)
 		}
 	}
 	for _, result := range results {
-		for _, hg := range result.hist {
-			err = hg.SaveToStorage(e.ctx, result.tableID, result.count, result.isIndex)
+		for _, hg := range result.Hist {
+			err = hg.SaveToStorage(e.ctx, result.TableID, result.Count, result.IsIndex)
 			if err != nil {
 				return nil, errors.Trace(err)
 			}
 		}
 	}
-	dom := sessionctx.GetDomain(e.ctx)
-	lease := dom.StatsHandle().Lease
-	if lease > 0 {
-		// We sleep two lease to make sure other tidb node has updated this node.
-		time.Sleep(lease * 2)
-	} else {
-		err := dom.StatsHandle().Update(GetInfoSchema(e.ctx))
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
+	err = dom.StatsHandle().Update(GetInfoSchema(e.ctx))
```

Comment:
Only when stats lease equal 0, it is called in different places, but in this case,  it  is normally in  test mode.